### ISSUE TITLE
using loss optimiser alongside swa

### ISSUE THREAD

**Issue by abaqusstudent** [KevinMusgrave/pytorch-metric-learning#469]
Hi,

I am using swa in pytorch via SWALR and so far I only need the model to be SWAed so the default code is fine when using with say for example arcface loss like so:

```
from torch.optim.swa_utils import AveragedModel, SWALR
from torch.optim.lr_scheduler import CosineAnnealingLR

loader, optimizer, model, loss_fn = ...
swa_model = AveragedModel(model)
scheduler = CosineAnnealingLR(optimizer, T_max=100)
swa_start = 5
swa_scheduler = SWALR(optimizer, swa_lr=0.05)

for epoch in range(100):
      for input, target in loader:
          optimizer.zero_grad()
          loss_fn(model(input), target).backward()
          optimizer.step()
      if epoch > swa_start:
          swa_model.update_parameters(model)
          swa_scheduler.step()
      else:
          scheduler.step()

# Update bn statistics for the swa_model at the end
torch.optim.swa_utils.update_bn(loader, swa_model)
# Use swa_model to make predictions on test data 
preds = swa_model(test_input)
```

however, now, I would like to use subcenter arcface loss and I am wondering if I need to SWA the loss_func aswell, so that I can compute the dominant centers at the end?

So, along the lines of (modifications in bold):

```
from torch.optim.swa_utils import AveragedModel, SWALR
from torch.optim.lr_scheduler import CosineAnnealingLR

loader, optimizer, model, loss_fn = ...
swa_model = AveragedModel(model)
**swa_af_loss_func = AveragedModel(loss_fn) ## subcenter AF loss.**
scheduler = CosineAnnealingLR(optimizer, T_max=100)
swa_start = 5
swa_scheduler = SWALR(optimizer, swa_lr=0.05)

for epoch in range(100):
      for input, target in loader:
          optimizer.zero_grad()
          loss_fn(model(input), target).backward()
          optimizer.step()
      if epoch > swa_start:
          swa_model.update_parameters(model)
          **swa_af_loss_func.update_parameters(loss_fn) ## average the loss func params**
          swa_scheduler.step()
      else:
          scheduler.step()

# Update bn statistics for the swa_model at the end
torch.optim.swa_utils.update_bn(loader, swa_model)
**torch.optim.swa_utils.update_bn(loader, swa_af_loss_func) #update bn for loss func**

# Use swa_model to make predictions on test data 
preds = swa_model(test_input)
**outliers, dominant_centers = swa_af_loss_func.get_outliers(
    embeddings, labels, threshold=75, return_dominant_centers=True
)**
```

Does this make sense?

Thank you for all your help.

**Comment by KevinMusgrave** [comment]
I think your `swa_af_loss_func` code makes sense, but I don't know if there's an advantage to using SWA for Subcenter ArcFace.
